generator/mavgen_java.py

```diff
@@ -364,24 +373,47 @@ def generate_MAVLinkMessage(directory, xml_list):
 
         payload.resetIndex();
 
-        final int payloadSize = payload.size();
-        for (int i = 0; i < payloadSize; i++) {
+        for (int i = 0; i < length; i++) {
             crc.update_checksum(payload.getByte());
         }
         crc.finish_checksum(msgid);
     }
 
+    /**
+     * Return length of actual data after triming zeros at the end.
+     * @param payload
+     * @return minimum length of valid data
+     */
+    private int mavTrimPayload (final MAVLinkPayload payload)
+    {
+
+        byte[] array = payload.payload.array();
+        int offset = payload.payload.position();
+
+        for (int i= offset-1; i>=0;--i)
+        {
+            if (array[i] != 0)
+            {
+                return i+1; 
+            }
+        }
+
+        return len;
+    }
+    
+
     /**
     * Encode this packet for transmission.
     *
     * @return Array with bytes to be transmitted
     */
     public byte[] encodePacket() {
-        byte[] buffer = new byte[10 + len + 2];
+        final int lenTrimmed = mavTrimPayload (payload);
```

Comment:
**lenTrimmed = mavTrimPayload (payload);**   is used instead of len in **encodePacket()**